### ISSUE TITLE
RFC compliance issues of parent proxies in transparent mode: request target of parent connect+ proxy request is not in "authorative form" in combination with transparent mode

### ISSUE THREAD

**Issue by JJ-Author** [3proxy/3proxy#1084]
I need to setup a transparent (for clients) proxy in front of an existing http and https via connect proxy (I have no further control over that)
A common task one would think, but I tried many proxy solutions - a lot of them fail. 
With your cool project I got very close to a working solution, so thank you for that.

However the connect+ upstream/parent proxy module seems to violate RFC 7231, by using the IP address in the CONNECT instead of the hostname used in the http request URI, when used as parent proxy in transparent mode scenario.
So when a client wants to request https://example.org/ which is transparently redirected with iptables to 3proxy the following CONNECT request is being formed by 3proxy to the parent proxy:

```
CONNECT 93.184.215.14:443 HTTP/1.0\r
Connection: keep-alive\r
\r
```
However, as per RFC a client (so 3proxy in this case) MUST send only the **target URI's authority** component [..] as the request-target.  (example.org:443 in this case).
As a result the upstream proxy refuses to serve the request for URLs that do not explicitly use the IP address (in other words refuses all DNS name based URLs when the CONNECT authority is using the IP, since only then the authority in request url == authority in CONNECT )

I think 3proxys misbehavior is probably rooted in the problem that the hostname for the parent proxy is assumed to be the original destination IP of the transparently redirected request (probably caused in the transparent plugin) but actually it has to be retrieved from the SNI field for https connection and from the http request itself for http requests. 
If I do not misunderstand transparent proxying of "modern" http(s) requests (http >= 1.1 for plaintext and https using SNI) the original IP destination is of no value since the (parent) proxy will try to resolve that anyhow.

Unfortunately the proxies I found which do that right, do not support a parent proxy via CONNECT.
So it would be pretty good to fix that behavior and maybe call it httpx-transparent or add some virtual parent proxy type that could be plugged in between transparent and actual parent.

NOTE: Although I am personally affected most by the problem with connect+ I also found that the correctness of other parent proxy modes seems affected in transparent mode:
- When it comes to socks5+ the connect message uses "address type. Domain name" but then passes analog to the http CONNECT mode the IP address as "remote name" - I did not read the SOCKS spec but I assume that is also not correct. 
- When using an http parent proxy the request line after the GET sent to upstream  http proxy is not in absolute form which is another violation and leads to an http 400 response code.

W.r.t. SNI I found a recent [commit](https://github.com/3proxy/3proxy/commit/013d4bc3339d8e3760855eae448972f3f34cba14), but I did not find enough documentation for that new tlspr and also the docker container dev image is not updated so I could not try it out (**would really appreciate if you could update the dev docker image**)
But maybe that code could be used to quickly apply a fix or already help in case of connect request?

<details><summary>my config</summary>
<p>

```
plugin /etc/3proxy/libexec/TransparentPlugin.ld.so transparent_plugin
log
logformat "L[%d-%m-%Y]: %U@%N:%p src: %C:%c dst: %R:%r orig_dst: %Q:%q ext_iface: %e hops: %h text: %T host: %n bytesI/0: %I/%O error: %E"                                    
auth iponly                                                                            
allow *                                  
parent 1000 connect+ proxy 8897 user1 mypassword                           
transparent
tcppm -i0.0.0.0 8081 127.0.0.1 11111   
```

</p>
</details> 

<details><summary>RFC quotations</summary>
<p>

>    A client sending a CONNECT request MUST send the **authority form** of
>    **request-target** (Section 5.3 of [RFC7230]); i.e., the request-target
>    consists of only the host name and port number of the tunnel
>    destination, separated by a colon.  For example,
> 
>      CONNECT server.example.com:80 HTTP/1.1
>      Host: server.example.com:80

> 5.3.3.  authority-form
> 
>    The authority-form of request-target is only used for CONNECT
>    requests (Section 4.3.6 of [RFC7231]).
> 
>      authority-form = authority
> 
>    When making a CONNECT request to establish a tunnel through one or
>    more proxies, a client MUST send only the **target URI's authority**
>    component (excluding any userinfo and its "@" delimiter) as the
>    request-target.  For example,
> 
>      CONNECT www.example.com:80 HTTP/1.1

</p>
</details>

**Comment by z3APA3A** [comment]
In the case of transparent redirection proxy is not aware of hostname and only IP address is known. Actually, it does not violate any RFC, because https://1.2.3.4/ is a valid URI, so prohibition to use IP address is a specific of parent proxy's configuration, not an RFC requirement.
You can use tlspr with latest master version, but there is no released 3proxy version with tlspr yet.

**Comment by JJ-Author** [comment]
thanks for your answer. I think there seems some misunderstanding of what I wrote and potentially the RFCs as well.
I provided exact quotations where I see that violation in the bottom of my first message (w.r.t. IP-based URIs) please read them again. 
If you think I misinterpreted that RFC then please explain.

With respect to the absolute-form requirement of the request-target I did not include quotes because I thought they would be more well-known to anybody dealing with proxy implementations but I will add here the quote from RFC 7230

```
5.3.2.  absolute-form

   When making a request to a proxy, other than a CONNECT or server-wide
   OPTIONS request (as detailed below), a client MUST send the target
   URI in absolute-form as the request-target.

     absolute-form  = absolute-URI

   The proxy is requested to either service that request from a valid
   cache, if possible, or make the same request on the client's behalf
   to either the next inbound proxy server or directly to the origin
   server indicated by the request-target.  Requirements on such
   "forwarding" of messages are defined in Section 5.7.

   An example absolute-form of request-line would be:

     GET http://www.example.org/pub/WWW/TheProject.html HTTP/1.1

```


NOTE:  that I never said something about an invalid URL. It`s just not valid/correct to be sent as request-target **to an upstream proxy** 

NOTE: I am not talking about a specific proxy (although I tried of course) I am talking about a general proxy that behaves according to the RFC as it is recommended for a correct implementation. 


Given your answer I deduct that using transparent proxy plugin with parent http and https via connect upstream and socks5+ upstream proxy are not a general valid configuration and only work for some use cases.

It makes sense to me that you state those limitations in the documentation and label this issue as wontfix (question label seems not appropriate since I do not really have question that was answered here ;-))

I think the bug with the "socks parent proxy wrong address field type" is probably easy to fix though (since you just need to set it to the proper 0x01 for IPv4  and 0x04 value for IPv6 addresses instead of 0x03 for domain-names (which clearly and IP address not is ;-)) 

I got tlspr working with transparent plugin and transparent keyword, and it seems to work and also solves that
problem with the invalid CONNECT request-target.  There seems one bug - the port flag does not seem to have an effect. I don't recall the details but I noted down in [my docker compose plug-and-play transparent proxying setup](https://github.com/dbpedia/ontology-time-machine/blob/28ea2e4f3125d8bf6604c63bedba890fe05e23f6/tests/transparent-setup/3proxy-TLS.cfg) 
443 does not seem to have an effect neither when using transparent plugin (which is good) nor when not using it without it (then weird target port is used but not 443)

Surprisingly I found for http proxying you actually don`t need the transparent proxy with tcppm you can just forward the request via iptables rules directly to it and then the problem with the request-target not being in absolute form is just gone. 

**SUMMARY/CONCLUSION**
So it is really just needs to be made clear in your documentation when the transparent plugin makes sense and when not. How I configured it initially as written in the first message has many unintended side effects that make that setup unreliable and dysfunctional for strict and less failure-tolerant upstream proxies.
I also would propose to add in the docu (of the plugin) how you actually create a proper transparent proxying setup for http and https traffic. W.r.t. to latter - everybody who reads this here and wonders how to do it you can have a look at [my docker compose plug-and-play transparent proxying setup](https://github.com/dbpedia/ontology-time-machine/blob/28ea2e4f3125d8bf6604c63bedba890fe05e23f6/tests/transparent-setup/3proxy-TLS.cfg) it is still work in progress and not fully tested but already seems quite nice thanks to 3proxy and docker.

**Comment by z3APA3A** [comment]
See

https://datatracker.ietf.org/doc/html/rfc7231#section-4.3.6
```

   A client sending a CONNECT request MUST send the authority form of
   request-target ([Section 5.3 of [RFC7230]](https://datatracker.ietf.org/doc/html/rfc7230#section-5.3));
```

https://datatracker.ietf.org/doc/html/rfc7230#section-5.3.3

`authority-form = authority
`

https://datatracker.ietf.org/doc/html/rfc7230#section-2.7

`authority     = <authority, see [[RFC3986], Section 3.2](https://datatracker.ietf.org/doc/html/rfc3986#section-3.2)>
`

https://datatracker.ietf.org/doc/html/rfc3986#section-3.2

`      authority   = [ userinfo "@" ] host [ ":" port ]
`
https://datatracker.ietf.org/doc/html/rfc3986#section-3.2.2

```
   The host subcomponent of authority is identified by an IP literal
   encapsulated within square brackets, an IPv4 address in dotted-
   decimal form, or a registered name.
```

**Comment by JJ-Author** [comment]
ok thank you for the detailed thought flow. I think now I understand where you are coming from. Unfortunately, I am not a native English speaker so maybe I did not write what I mean very well. 
I will try again with the grammar rules you posted and this additional one:

`host        = IP-literal / IPv4address / reg-name` shows that an IP(v4) address is a valid `host` and therefore a valid `authority` and therefore a valid "general " `authority-form`. I think we both agree on that. 

but when the request-target of a request that is supposed to be sent to an https proxy via CONNECT is: 

`https://foo@example.org:441/bar` 

I understand that 
`authority   = [ username "@" ] host [ ":" 411]`

`host = regname = example.org`

`username = foo`

however `authority-form` for CONNECT  is `example.org:441` since the username must be pruned from authority according to the spec. 
3proxy however would provide in the case described `93.184.215.14:441` which is to me a valid `authority-form` **but not of the request-target**. it  would be the correct `authority-form` for `https://foo@93.184.215.14:441/bar` , but 
those are 2 different request-targets with 2 different `authorities`.

I understand why it is easier in transparent mode to do that, but it doesnt seem to change the fact the plugin documentation is misleading and in my opinion indicates something that is not achievable with the current implementation for http and https protocols 

> Plugin can turn 3proxy into transparent proxy for **virtually any TCP-based protocol** and use **all 3proxy features** - redirections, **parent proxies**, ACLs, traffic limitations, etc. TransparentPlugin plugin takes destination IP:port from Linux and uses this information as a **target IP in proxy**.

Since I know the details I got a theory what you wanna say with that. 
But when combining with parent proxy chaining, the target IP is only the request-target-host for those requests that already use the IP as host which nowadays seems not to be the case for most https request-targets.
